main/src/cgeo/geocaching/maps/routing/Routing.java

```diff
@@ -207,6 +207,14 @@ public static synchronized void disconnect(final String callbackKey) {
             return null;
         }
 
+        // wanted to compare (gpx.substr(0,5) != "<?xml")
+        // but this did not work, most likely because gpx is a Unicode string?
+
+        if (gpx.length() < 200) {
```

Comment:
I think comparing the length isn't a good idea. what about `!gpx.startsWith("<?xml")`?

Reply:
Yes, I agree, it is not a good idea, however, if this: `(gpx.substr(0,5) != "<?xml")` did not work, and it didn't, why would that: `!gpx.startsWith("<?xml")` work any better? ... Well ... I will just go ahead and try it.

Reply:
Why it is not working exactly? Have you tried to debug what is returned by `gpx.substr(0,5)` or is it raising any error? I don't really understand why it shouldn't work...

Reply:
I had put an unconditional: `Log.w(gpx.substr(0,5))` in front of the if () statement and as expected, it logged `<?xml` but the if () expression never evaluaet to `true`. I played around with substr(0,4) vs. substr(0,5) but nothing ever made it a `true` so as an act of some sort of desparation I coded the .lenth() thing and it worked right away. I tested both paths, one that returned a valid XML and one that returned an error message.

Reply:
I now tried: `!gpx.startsWith("<?xml")` ... and it worked. Perhaps I should go back and try my initial `(gpx.substr(0,5) != "<?xml")` for completeness.....

Reply:
Again, didn't work! Why? Beats me. I use your recommendation gpx.startsWith and will push this now.

Reply:
I guess this is due to the way Java compares strings: If you use "==" comparison operator on strings, only the object addresses will be compared, not the content. To compare the strings' contents, you need method like `string1.equals(string2)` or the helper methods from `StringUtil` like `StringUtil.isNotBlank(string1)` etc.

Reply:
Thank you very  much. This explanation resonates well. As you can clearly see, I am new to Java. Something like this would perfectly work in JavaScript.

On the side: I am actually surprised that Google hasn't made the move to put all Android Apps on top of their V8 engine. It would signifiantly simplify the ecosystem as this is what runs in the Browsers as well as in serverside Node.JS which Google is a big fan of for both.